Source/Core/InputCommon/ControllerInterface/WGInput/WGInput.cpp

```diff
@@ -0,0 +1,686 @@
+// Copyright 2020 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "InputCommon/ControllerInterface/WGInput/WGInput.h"
+
+#include <array>
+#include <string_view>
+
+#include <fmt/format.h>
+
+#include <roapi.h>
+#include <windows.gaming.input.h>
+#include <wrl.h>
+
+#include "Common/Logging/Log.h"
+#include "Common/StringUtil.h"
+#include "InputCommon/ControllerInterface/ControllerInterface.h"
+
+#pragma comment(lib, "runtimeobject.lib")
+
+namespace WGI = ABI::Windows::Gaming::Input;
+using ABI::Windows::Foundation::Collections::IVectorView;
+using Microsoft::WRL::ComPtr;
+
+namespace
+{
+bool g_runtime_initialized = false;
+bool g_runtime_needs_deinit = false;
+}  // namespace
+
+namespace ciface::WGInput
+{
+static constexpr std::string_view SOURCE_NAME = "WGInput";
+
+// These names correspond to the values of the GameControllerButtonLabel enum.
+// "None" is not used.
+// There are some overlapping names assuming no device exposes both
+// GameControllerButtonLabel_XboxLeftBumper and GameControllerButtonLabel_LeftBumper.
+// If needed we can prepend "Xbox" to relevant input names on conflict in the future.
+static constexpr std::array wgi_button_names = {
+    "None",      "Back",     "Start",      "Menu",     "View",     "Pad N",
+    "Pad S",     "Pad W",    "Pad E",      "Button A", "Button B", "Button X",
+    "Button Y",  "Bumper L", "Trigger L",  "Thumb L",  "Bumper R", "Trigger R",
+    "Thumb R",   "Paddle 1", "Paddle 2",   "Paddle 3", "Paddle 4", "Mode",
+    "Select",    "Menu",     "View",       "Back",     "Start",    "Options",
+    "Share",     "Pad N",    "Pad S",      "Pad W",    "Pad E",    "Letter A",
+    "Letter B",  "Letter C", "Letter L",   "Letter R", "Letter X", "Letter Y",
+    "Letter Z",  "Cross",    "Circle",     "Square",   "Triangle", "Bumper L",
+    "Trigger L", "Thumb L",  "Left 1",     "Left 2",   "Left 3",   "Bumper R",
+    "Trigger R", "Thumb R",  "Right 1",    "Right 2",  "Right 3",  "Paddle 1",
+    "Paddle 2",  "Paddle 3", "Paddle 4",   "Plus",     "Minus",    "Down Left Arrow",
+    "Dial L",    "Dial R",   "Suspension",
+};
+
+static constexpr std::array gamepad_trigger_names = {"Trigger L", "Trigger R"};
+static constexpr std::array gamepad_axis_names = {"Left X", "Left Y", "Right X", "Right Y"};
+static constexpr std::array gamepad_motor_names = {"Motor L", "Motor R", "Trigger L", "Trigger R"};
+
+class Device : public Core::Device
+{
+public:
+  Device(std::string name, ComPtr<WGI::IRawGameController> raw_controller, WGI::IGamepad* gamepad)
+      : m_name(std::move(name)), m_raw_controller(raw_controller), m_gamepad(gamepad)
+  {
+    // Buttons:
+    PopulateButtons();
+
+    // Add inputs for IGamepad interface if available.
+    if (m_gamepad)
+    {
+      // Axes:
+      const double* axis = &m_gamepad_reading.LeftThumbstickX;
```

Comment:
This here is spooky, even if it technically works. I'd rather this was more explicit, maybe something along the lines of

~~~
      const std::array axis_ptrs{
          &m_gamepad_reading.LeftThumbstickX, &m_gamepad_reading.LeftThumbstickY,
          &m_gamepad_reading.RightThumbstickX, &m_gamepad_reading.RightThumbstickY};
      size_t axis = 0;
      for (auto& axis_name : gamepad_axis_names)
      {
        AddInput(new NamedAxis(axis_ptrs[axis], 0.0, -1.0, axis_name));
        AddInput(new NamedAxis(axis_ptrs[axis], 0.0, +1.0, axis_name));
        ++axis;
      }
~~~

Likewise for the triggers and motors below.

Reply:
Fixed.